src/Common/StringParsing.fs

```diff
@@ -14,44 +14,57 @@ open FSharp.Collections
 
 module String =
   /// Matches when a string is a whitespace or null
-  let (|WhiteSpace|_|) s = 
+  let (|WhiteSpaceS|_|) (s) = 
+    if String.IsNullOrWhiteSpace(s) then Some() else None
+
+  /// Matches when a string is a whitespace or null
+  let (|WhiteSpace|_|) (s, n: int) = 
     if String.IsNullOrWhiteSpace(s) then Some() else None
 
   /// Matches when a string does starts with non-whitespace
-  let (|Unindented|_|) (s:string) = 
+  let (|Unindented|_|) (s:string, n:int) = 
     if not (String.IsNullOrWhiteSpace(s)) && s.TrimStart() = s then Some() else None
 
   /// Returns a string trimmed from both start and end
-  let (|TrimBoth|) (text:string) = text.Trim()
+  let (|TrimBothS|) (text:string) = text.Trim()
+  /// Returns a string trimmed from both start and end
+  let (|TrimBoth|) (text:string, n:int) = (text.Trim(), n)
   /// Returns a string trimmed from the end
-  let (|TrimEnd|) (text:string) = text.TrimEnd()
+  let (|TrimEnd|) (text:string, n:int) = (text.TrimEnd(), n)
   /// Returns a string trimmed from the start
-  let (|TrimStart|) (text:string) = text.TrimStart()
+  let (|TrimStart|) (text:string, n:int) = (text.TrimStart(), n)
 
   /// Retrusn a string trimmed from the end using characters given as a parameter
-  let (|TrimEndUsing|) chars (text:string) = text.TrimEnd(Array.ofSeq chars)
+  let (|TrimEndUsing|) chars (text:string, n:int) = text.TrimEnd(Array.ofSeq chars)
 
   /// Returns a string trimmed from the start together with 
   /// the number of skipped whitespace characters
-  let (|TrimStartAndCount|) (text:string) = 
+  let (|TrimStartAndCount|) (text:string, n:int) = 
     let trimmed = text.TrimStart([|' '; '\t'|])
     let len = text.Length - trimmed.Length
-    len, text.Substring(0, len).Replace("\t", "    ").Length, trimmed
+    len, text.Substring(0, len).Replace("\t", "    ").Length, (trimmed, n)
 
   /// Matches when a string starts with any of the specified sub-strings
-  let (|StartsWithAny|_|) (starts:seq<string>) (text:string) = 
+  let (|StartsWithAny|_|) (starts:seq<string>) (text:string, n:int) = 
     if starts |> Seq.exists (text.StartsWith) then Some() else None
   /// Matches when a string starts with the specified sub-string
-  let (|StartsWith|_|) (start:string) (text:string) = 
+  let (|StartsWithS|_|) (start:string) (text:string) = 
```

Comment:
I think the `S` suffix on the active patterns is not very idiomatic. Could we instead have two modules, say `String` and `StringPosition` so that the patterns that work on string with position are all in `StringPosition` module?